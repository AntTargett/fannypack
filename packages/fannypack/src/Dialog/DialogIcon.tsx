import * as React from 'react';
import * as PropTypes from 'prop-types';

import { IconProps } from '../Icon/Icon';
import { DialogIcon as _DialogIcon } from './styled';

export type LocalDialogIconProps = {
  className?: string;
  icon: string;
};
export type DialogIconProps = IconProps & LocalDialogIconProps;

// @ts-ignore
export const DialogIcon: React.FunctionComponent<LocalDialogIconProps> = props => <_DialogIcon {...props} />;

export const dialogIconPropTypes = {
  className: PropTypes.string,
  icon: PropTypes.string.isRequired
};
DialogIcon.propTypes = dialogIconPropTypes;

export const dialogIconDefaultProps = {
  className: undefined
};
DialogIcon.defaultProps = dialogIconDefaultProps;

// @ts-ignore
const C: React.FunctionComponent<DialogIconProps> = DialogIcon;
export default C;
