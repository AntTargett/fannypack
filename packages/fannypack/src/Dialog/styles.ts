import { css, cssClass } from '../styled';
import { borderRadius, fontSize, fontWeight, palette, space, theme } from '../utils';

export const Dialog = styleProps => cssClass`
  border-radius: ${borderRadius('default')(styleProps)};

  ${theme('altitudes.100')(styleProps)};

  & {
    ${theme(`${styleProps.themeKey}.base`)(styleProps)};
  }
`;

export const DialogContent = styleProps => cssClass`
  padding: ${space(3, 'major')(styleProps)}rem;

  & {
    ${theme(`${styleProps.themeKey}.base`)(styleProps)};
  }
`;

export const DialogHeader = styleProps => cssClass`
  align-items: center;
  display: flex;
  justify-content: space-between;
  border-bottom: 1px solid ${palette('white800')(styleProps)};
  padding: ${space(5, 'minor')(styleProps)}rem ${space(6, 'minor')(styleProps)}rem;

  & {
    ${theme(`${styleProps.themeKey}.base`)(styleProps)};
  }
`;

export const DialogTitle = styleProps => cssClass`
  font-size: ${fontSize('300')(styleProps)}em;
  font-weight: ${fontWeight('semibold')(styleProps)};

  & {
    ${theme(`${styleProps.themeKey}.base`)(styleProps)};
  }
`;

export const DialogFooter = styleProps => cssClass`
  align-items: center;
  border-top: 1px solid ${palette('white800')(styleProps)};
  display: flex;
  padding: ${space(5, 'minor')(styleProps)}rem ${space(6, 'minor')(styleProps)}rem;

  ${styleProps.showActionButtons &&
    styleProps.footer &&
    css`
      justify-content: space-between;
    `};

  ${styleProps.showActionButtons &&
    !styleProps.footer &&
    css`
      justify-content: flex-end;
    `};

  & {
    ${theme(`${styleProps.themeKey}.base`)(styleProps)};
  }
`;

export const DialogClose = styleProps => cssClass`
  & {
    ${theme(`${styleProps.themeKey}.base`)(styleProps)};
  }
`;

export const DialogIconWrapper = styleProps => cssClass`
  margin-right: ${space(2, 'major')(styleProps)}rem;

  & {
    ${theme(`${styleProps.themeKey}.base`)(styleProps)};
  }
`;
