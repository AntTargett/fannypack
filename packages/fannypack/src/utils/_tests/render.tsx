import * as React from 'react';
import { render } from '@testing-library/react';
// @ts-ignore
import { Provider } from '../../Provider';

export default (Component: any, { theme = {} } = {}) => render(<Provider theme={theme}>{Component}</Provider>);
