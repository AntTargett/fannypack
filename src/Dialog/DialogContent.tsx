import * as React from 'react';
import * as PropTypes from 'prop-types';
import { BoxProps as ReakitBoxProps } from '@jmoxey/reakit/ts';

import { DialogContent as _DialogContent } from './styled';

export type LocalDialogContentProps = {
  children: React.ReactNode;
  className?: string;
};
export type DialogContentProps = LocalDialogContentProps & ReakitBoxProps;

export const DialogContent: React.FunctionComponent<LocalDialogContentProps> = ({ children, ...props }) => (
  <_DialogContent {...props}>{children}</_DialogContent>
);

DialogContent.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string
};
DialogContent.defaultProps = {
  className: undefined
};

const C: React.FunctionComponent<DialogContentProps> = DialogContent;
export default C;
