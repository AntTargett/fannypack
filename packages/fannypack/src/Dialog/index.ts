import Dialog from './Dialog';
import * as styledDialog from './styled';

export default Dialog;

export * from './Dialog';
export { default as Dialog } from './Dialog';
export * from './DialogContent';
export { default as DialogContent } from './DialogContent';
export * from './DialogClose';
export { default as DialogClose } from './DialogClose';
export * from './DialogDialog';
export { default as DialogDialog } from './DialogDialog';
export * from './DialogFooter';
export { default as DialogFooter } from './DialogFooter';
export * from './DialogHeader';
export { default as DialogHeader } from './DialogHeader';
export * from './DialogTitle';
export { default as DialogTitle } from './DialogTitle';
export * from './DialogIcon';
export { default as DialogIcon } from './DialogIcon';

export { styledDialog };
