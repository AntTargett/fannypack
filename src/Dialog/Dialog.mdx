---
name: Dialog
route: /components/Dialog
menu: Components
---

import { Playground, PropsTable } from 'docz';
import { Box } from '../primitives/index';
import { ActionButtons } from '../Button';
import Heading from '../Heading';
import Paragraph from '../Paragraph';
import Set from '../Set';
import Dialog from './index';

# Dialog

## Import

`import { Dialog } from 'fannypack'`

## Basic Usage

<Playground>
  <Dialog>
    <Dialog.Header>
      <Dialog.Title>This is a title</Dialog.Title>
    </Dialog.Header>
    <Dialog.Content>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse diam ipsum, cursus id placerat congue,
      ultrices eget lectus. Duis posuere, lacus sed tristique commodo, sapien turpis mollis nunc, vestibulum consectetur
      lectus augue sit amet justo.
    </Dialog.Content>
  </Dialog>
</Playground>

## Footer

You can render a footer for the dialog using the `footer` prop.

<Playground>
  <Dialog>
    <Dialog.Header>
      <Dialog.Title>This is a title</Dialog.Title>
    </Dialog.Header>
    <Dialog.Content>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse diam ipsum, cursus id placerat congue,
      ultrices eget lectus. Duis posuere, lacus sed tristique commodo, sapien turpis mollis nunc, vestibulum consectetur
      lectus augue sit amet justo.
    </Dialog.Content>
    <Dialog.Footer>
      <ActionButtons />
    </Dialog.Footer>
  </Dialog>
</Playground>

## Props

<PropsTable of={Dialog} />

## Theming

### Schema

```jsx
{
  base: string | Object,
  Header: {
    base: string | Object
  },
  Content: {
    base: string | Object
  },
  Footer: {
    base: string | Object
  }
}
```
