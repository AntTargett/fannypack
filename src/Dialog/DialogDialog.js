// @flow
import React, { type Node } from 'react';

import Dialog from './styled';

type Props = {
  a11yDescriptionId?: string,
  a11yTitleId?: string,
  border?: true | 'shadow',
  children: Node,
  className?: string
};

const DialogDialog = ({ a11yDescriptionId, a11yTitleId, border, children, ...props }: Props) => (
  <Dialog aria-describedby={a11yDescriptionId} aria-labelledby={a11yTitleId} role="dialog" border={border} {...props}>
    {children}
  </Dialog>
);

DialogDialog.defaultProps = {
  a11yDescriptionId: undefined,
  a11yTitleId: undefined,
  border: true,
  className: undefined
};

export default DialogDialog;
