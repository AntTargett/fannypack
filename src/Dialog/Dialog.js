// @flow
import React, { type Element, type Node } from 'react';

import { getUniqueId } from '../uniqueId';
import { Box } from '../primitives';
import ActionButtons, { type Props as ActionButtonsProps } from '../Button/ActionButtons';
import DialogDialog from './DialogDialog';
import DialogContent from './DialogContent';
import DialogHeader from './DialogHeader';
import DialogFooter from './DialogFooter';
import DialogTitle from './DialogTitle';
import DialogClose from './DialogClose';
import DialogIcon from './DialogIcon';

type Props = {
  actionButtonsProps?: ActionButtonsProps,
  a11yDescriptionId?: string,
  a11yTitleId?: string,
  border?: true | 'shadow',
  children: Node,
  className?: string,
  closeButtonProps?: Object,
  footer?: string | Element<any>,
  kind?: ?'alert',
  onClickClose?: Function,
  showActionButtons?: boolean,
  showCloseButton?: boolean,
  title?: string | Element<any>,
  type?: string
};

const Dialog = ({
  actionButtonsProps,
  a11yDescriptionId,
  a11yTitleId,
  border,
  children,
  closeButtonProps,
  footer,
  kind,
  onClickClose,
  showActionButtons,
  showCloseButton,
  title,
  type,
  ...props
}: Props) => (
  <DialogDialog a11yDescriptionId={a11yDescriptionId} a11yTitleId={a11yTitleId} kind={kind} border={border} {...props}>
    {title && (
      <DialogHeader>
        {typeof title === 'string' ? (
          <DialogTitle id={a11yTitleId}>
            {type && <DialogIcon color={type} icon={type} />}
            {title}
          </DialogTitle>
        ) : (
          title
        )}
        {showCloseButton && <DialogClose onClick={onClickClose} {...closeButtonProps} />}
      </DialogHeader>
    )}
    <DialogContent id={a11yDescriptionId}>{children}</DialogContent>
    {(footer || showActionButtons) && (
      <DialogFooter justifyContent={footer ? 'space-between' : 'flex-end'}>
        {footer && <Box>{footer}</Box>}
        {showActionButtons && (
          <Box>
            <ActionButtons {...actionButtonsProps} />
          </Box>
        )}
      </DialogFooter>
    )}
  </DialogDialog>
);

Dialog.Dialog = DialogDialog;
Dialog.Header = DialogHeader;
Dialog.Content = DialogContent;
Dialog.Footer = DialogFooter;
Dialog.Title = DialogTitle;
Dialog.Close = DialogClose;
Dialog.Icon = DialogIcon;

Dialog.defaultProps = {
  actionButtonsProps: {},
  a11yDescriptionId: getUniqueId('Dialog'),
  a11yTitleId: getUniqueId('Dialog'),
  border: true,
  className: undefined,
  closeButtonProps: {},
  footer: undefined,
  kind: undefined,
  onClickClose: undefined,
  showActionButtons: false,
  showCloseButton: false,
  title: undefined,
  type: undefined
};

export default Dialog;
