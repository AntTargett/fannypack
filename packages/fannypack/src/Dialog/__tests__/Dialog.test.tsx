import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Dialog } from '../Dialog';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(<Dialog>Hello world</Dialog>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(<Dialog color="primary">Hello world</Dialog>);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(<Dialog use="div">Hello world</Dialog>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with Dialog props', () => {
      const { result } = renderHook(() => Dialog.useProps());
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(<Dialog>{DialogProps => <div {...DialogProps}>Hello world</div>}</Dialog>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('theming', () => {
  it('Dialog.base should render correctly', () => {
    const { container } = render(<Dialog>hello world</Dialog>, {
      theme: { Dialog: { base: { backgroundColor: 'red' } } }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(<Dialog>hello world</Dialog>, {
      theme: { Dialog: { defaultProps: { className: 'test', color: 'primary' } } }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
