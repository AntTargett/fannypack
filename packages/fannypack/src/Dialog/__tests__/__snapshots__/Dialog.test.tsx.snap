// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  border-radius: 4px;
  box-shadow: 0px 2px 4px 0px rgba(0,0,0,0.1),0px 0px 2px 1px rgba(0,0,0,0.02);
}

.emotion-0 {
  padding: 1.5rem;
}

<div
  aria-describedby="dialogDescription"
  class="emotion-3 emotion-1 emotion-2 fp-Box fp-Dialog"
>
  <div
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-Dialog.Content"
    id="dialogDescription"
  >
    Hello world
  </div>
</div>
`;

exports[`composition hook should return with Dialog props 1`] = `
Object {
  "aria-describedby": undefined,
  "aria-labelledby": undefined,
  "children": <Context.Provider
    value={
      Object {
        "descriptionId": "",
        "titleId": "",
      }
    }
  >
    <React.Fragment>
      <ForwardRef />
    </React.Fragment>
  </Context.Provider>,
  "className": "css-5r32jl css-1pwrqjj css-3ekj6a fp-Box fp-Dialog",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  border-radius: 4px;
  box-shadow: 0px 2px 4px 0px rgba(0,0,0,0.1),0px 0px 2px 1px rgba(0,0,0,0.02);
}

<div
  aria-describedby="dialogDescription"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Dialog"
>
  Hello world
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  border-radius: 4px;
  box-shadow: 0px 2px 4px 0px rgba(0,0,0,0.1),0px 0px 2px 1px rgba(0,0,0,0.02);
}

.emotion-0 {
  padding: 1.5rem;
}

.emotion-5 {
  color: #3722ed !important;
}

<div
  aria-describedby="dialogDescription"
  class="emotion-3 emotion-1 emotion-5 test fp-Box fp-Dialog"
>
  <div
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-Dialog.Content"
    id="dialogDescription"
  >
    hello world
  </div>
</div>
`;

exports[`props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  border-radius: 4px;
  box-shadow: 0px 2px 4px 0px rgba(0,0,0,0.1),0px 0px 2px 1px rgba(0,0,0,0.02);
}

.emotion-0 {
  padding: 1.5rem;
}

<div
  aria-describedby="dialogDescription"
  class="emotion-3 emotion-1 emotion-2 fp-Box fp-Dialog"
>
  <div
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-Dialog.Content"
    id="dialogDescription"
  >
    Hello world
  </div>
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  border-radius: 4px;
  box-shadow: 0px 2px 4px 0px rgba(0,0,0,0.1),0px 0px 2px 1px rgba(0,0,0,0.02);
}

.emotion-0 {
  padding: 1.5rem;
}

.emotion-5 {
  color: #3722ed !important;
}

<div
  aria-describedby="dialogDescription"
  class="emotion-3 emotion-1 emotion-5 fp-Box fp-Dialog"
>
  <div
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-Dialog.Content"
    id="dialogDescription"
  >
    Hello world
  </div>
</div>
`;

exports[`theming Dialog.base should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  padding: 1.5rem;
}

.emotion-3 {
  border-radius: 4px;
  box-shadow: 0px 2px 4px 0px rgba(0,0,0,0.1),0px 0px 2px 1px rgba(0,0,0,0.02);
}

.emotion-3 {
  background-color: red;
}

<div
  aria-describedby="dialogDescription"
  class="emotion-3 emotion-1 emotion-2 fp-Box fp-Dialog"
>
  <div
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-Dialog.Content"
    id="dialogDescription"
  >
    hello world
  </div>
</div>
`;
